modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/runner/TagHarvester.java

```diff
@@ -0,0 +1,118 @@
+
+import com.google.gson.JsonArray;
+
+import dev.galasa.Tags;
+import dev.galasa.framework.spi.DynamicStatusStoreException;
+import dev.galasa.framework.spi.IDynamicStatusStoreService;
+import dev.galasa.framework.spi.IResultArchiveStore;
+import dev.galasa.framework.spi.ResultArchiveStoreException;
+import dev.galasa.framework.spi.teststructure.TestStructure;
+import dev.galasa.framework.spi.utils.GalasaGson;
+
+/**
+ * Allows a test runner to harvest tags from a java class.
+ */
+public class TagHarvester {
+
+    Log logger = LogFactory.getLog(TagHarvester.class);
+
+    private IDynamicStatusStoreService dss;
+    private IResultArchiveStore ras;
+    private TestStructure testStructure;
+    private GalasaGson gson ;
+    public TagHarvester(IDynamicStatusStoreService dss, IResultArchiveStore ras, TestStructure testStructure, GalasaGson gson) {
+        this.dss = dss ;
+        this.ras = ras ;
+        this.testStructure = testStructure;
+        this.gson = gson;
+    }
+
+    /**
+     * We have a test class with possible annotations inside. 
+     * Harvest these and add them to the collection of tags we have from when the test was submitted.
+     * Then update the dss and ras accordingly.
+     * 
+     * @param testClass
+     */
+    public void harvestTagsFromTestClass( Class<?> testClass, String runName) {
+        Set<String> classAnnotationTags = extractTagsFromClass(testClass);
+        if( ! classAnnotationTags.isEmpty() ) {
+            // There were test class annotations which added tags.
+            updateTestStructureWithTags(classAnnotationTags);
+            updateTagsInDss(runName);
+            updateTagsInRas();
+        }
+    }
+
+    private Set<String> extractTagsFromClass(Class<?> clazz) {
+        HashSet<String> tags = new HashSet<String>();
+
+        Tags[] annotations = clazz.getAnnotationsByType(Tags.class);
+        if( annotations != null) {
+            for( Tags annotation : annotations ){
+                String[] values = annotation.value();
+                if( values != null) {
+                    for( String value : values) {
+                        if( value != null) {
+                            String trimmedValue = value.trim();
+                            if(!trimmedValue.isBlank() ){
+                                tags.add(trimmedValue);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return tags;
+    }
+
+    private void updateTestStructureWithTags(Set<String> tagsToAdd) {
+        // Update the test structure with tags.
+        for(String tag: tagsToAdd ) {
+            this.testStructure.addTag(tag);
+        }
+    }
+
+    private void updateTagsInDss(String runName) {
+
+        // Update the dss with the tags we have collected in the test structure.
+        Set<String> tags = this.testStructure.getTags();
+        String key = "run." + runName + "." + "tags" ;
```

Comment:
It's a shame that the IRun which reads these doesn't also offer a write-to-dss to group all the dss-manipulation. Maybe this should be moved.